### ISSUE TITLE
How to create a simulation grid with custom obstacles?

### ISSUE THREAD

**Issue by sambaPython24** [tum-pbs/PhiFlow#185]
Hey, thank you for your work and for your support on github!

In your examples pages, you show a nice example of how to create grids with custom obstacles, that are 
created as boolean operations of [boxes](https://github.com/tum-pbs/PhiFlow/blob/master/examples/mesh/FVM_Heat.ipynb) or how to load 
a [gmsh file](https://github.com/tum-pbs/PhiFlow/blob/master/examples/mesh/FVM_Cylinder_GMsh.ipynb) as a grid. 

Is there a more direct way of how to load a mesh for example by e.g. :
- Giving e.g. a boolean voxel mask of where an object is (e.g. True) or is not (False) or 
- Loading a mesh consisting of e.g. vertices and faces (like for .obj files) without having to save it as a .msh file. 

and - because I see only 2D Simulation examples in your code - 

is that also possible for 3D Simulations?

**Comment by holl-** [comment]
Hi @sambaPython24 Yes, 3D is also supported.
For obstacles on a regular grid, you can pass them directly to `make_incompressible()`. [Here](https://tum-pbs.github.io/PhiFlow/examples/grids/Batched_Smoke.html) is an example. Of course, you can also union them!

**Comment by sambaPython24** [comment]
Thank you @holl- for your quick answer and indeed I see the line
``fluid.make_incompressible(v, obstacle, Solve(x0=p))``` and that you have been creating obstacles using the predefined Cuboid and Spheres. In another tutorial, one could also upload gmsh file. 

I assume, that you convert the Cuboid and Spheres to a part of a voxel grid at some point anyways, hence my question is wether I could just skip this step and provide this voxel grid directly (either as the simulation grid or just a binary grid with the location of the "blocked" voxel).

**Comment by holl-** [comment]
The reason you provide a `Geometry` or `Obstacle` object instead of a voxel grid is that the distance to the obstacle surface is also used in the computation.
If you only have a voxel grid, you can look at the experimental class `phi.geom._voxels.Voxels`, but this may not work for your use case.

**Comment by sambaPython24** [comment]
@holl- 
But on a fundamental level, the geometric information, the form of the grid and its occupation must be either stored in ```torch.tensor``` somewhere or the underlying differential equation is evaluated not inside a grid but grid-free. 

Which option is true? And how do I find this ```torch.tensor```?

**Comment by holl-** [comment]
Yes, you're right. The obstacle is sampled on the simulation grid [here inside `make_incompressible`](https://github.com/tum-pbs/PhiFlow/blob/95243a7e7936da045e74e0fdc96d833fb2e7e097/phi/physics/fluid.py#L133)
You can get the underlying tensor from there, e.g. using `accessible.values.native('x,y')`. Note that this could be a NumPy array or PyTorch tensor, depending on the obstacle parameterization.
